v2/client/receiver_test.go

```diff
@@ -38,6 +39,9 @@ func TestReceiverFnValidTypes(t *testing.T) {
 		"ctx in, Event+Result out":       func(context.Context) (*event.Event, protocol.Result) { return nil, nil },
 		"Event in, Event+Result out":     func(event.Event) (*event.Event, protocol.Result) { return nil, nil },
 		"ctx+Event in, Event+Result out": func(context.Context, event.Event) (*event.Event, protocol.Result) { return nil, nil },
+
+		"input contravariance; may accept supertype": func(event.EventReader) {},
+		"output covariance; may return subtype": func() *myErr { return nil },
```

Comment:
I would not expect this to work. the `protocol.Result` is an interface and what it creates internally will not be cast-able to a `*myEr`r unless it was an interface.

Reply:
The `protocol.Result` interface is indeed not convertible to `*myErr`, it is the other way around. The `*myErr` type implements `protocol.Result` and therefore is convertible to it, and that's what is needed for `output types`. It works exactly opposite for `input types`.

Reply:
I am little surprised that `event.Event` (non-pointer type) is convertible to interface type `event.EventReader`, I totally understand that pointer type `*event.Event` is convertible `event.EventReader`.

Reply:
I tried to run:
```go
func Handle(ctx context.Context, event event.EventReader) {
	fmt.Println(event.ID())
}
```
and it worked. Not saying it's particularly useful.

Reply:
I think go does some magic to make non-pointers translate to interfaces when the pointer of that struct does implement the interface.

Reply:
>I think go does some magic to make non-pointers translate to interfaces when the pointer of that struct does implement the interface.

Seems it really does, but only if receivers of given interfaces are all non-pointer, interesting.